src/chain/chain_information/babe_current_epoch.rs

```diff
@@ -0,0 +1,168 @@
+// Substrate-lite
+// Copyright (C) 2019-2021  Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{
+    chain::chain_information::BabeEpochInformation,
+    executor::{host, read_only_runtime_host},
+    header,
+};
+
+use alloc::vec::Vec;
+use parity_scale_codec::{Decode, Encode};
+
+/// Configuration for [`babe_current_epoch`].
+pub struct Config {
+    /// Runtime used to get the current babe epoch. Must be built using the Wasm code found at the
+    /// `:code` key of the parent block storage.
+    pub parent_runtime: host::HostVmPrototype,
+}
+
+/// Problem encountered during a call to [`babe_current_epoch`].
+#[derive(Debug, Clone, derive_more::Display)]
+pub enum Error {
+    /// Error while starting the Wasm virtual machine.
+    #[display(fmt = "{}", _0)]
+    WasmStart(host::StartErr),
+    /// Error while running the Wasm virtual machine.
+    #[display(fmt = "{}", _0)]
+    WasmVm(read_only_runtime_host::Error),
+    /// Error while decoding the babe epoch.
+    #[display(fmt = "{}", _0)]
+    DecodeFailed(parity_scale_codec::Error),
+}
+
+/// Fetches the current Babe epoch using `BabeApi_current_epoch`.
+pub fn babe_current_epoch(config: Config) -> Query {
+    let vm = read_only_runtime_host::run(read_only_runtime_host::Config {
+        virtual_machine: config.parent_runtime,
+        function_to_call: "BabeApi_current_epoch",
+        // `BabeApi_current_epoch`doesn't take any parameters.
+        parameter: std::iter::empty::<&[u8]>(),
+    });
+
+    match vm {
+        Ok(vm) => Query::from_inner(vm),
+        Err(err) => Query::Finished(Err(Error::WasmStart(err))),
+    }
+}
+
+/// Current state of the operation.
+#[must_use]
+pub enum Query {
+    /// Fetching the Babe epoch is over.
+    Finished(Result<BabeEpochInformation, Error>),
```

Comment:
You also need to return the `HostVmPrototype`, as these are super expensive to create.

```suggestion
    Finished { result: Result<BabeEpochInformation, Error>, runtime: host::HostVmPrototype },
```

Reply:
We should change the return type of `Finished` in `read_only_runtime_host` then too.

Reply:
At the moment we don't have access to the runtime in the case of `read_only_runtime_host::RuntimeHostVm::Finished(Err(err)) ` or `read_only_runtime_host::run` returning `Err` (which is impossible at present).

Reply:
Yes. It's really important to keep this object alive.

Reply:
We can change `RuntimeHostVm::Finished` to:
```rust
Finished(Result<Success, (Error, host::HostVmPrototype)>)
```

by adding an `into_prototype()` function to `host::HostVm` variants, including forbidden ones and calling that.

Getting `read_only_runtime_host::run` to return the prototype on `Err` is harder though, as it would require returning a `HostVmPrototype` in `HostVmPrototype::run_vectored` and that requires going into the individual VM implementations and getting them to return `Self` on error as well.